### PR TITLE
KSM core: add a `kubernetes_state.cronjob.complete` service check

### PR DESCRIPTION
**Pull request by L3n41c**
### What does this PR do?

Add a new service check in the KSM core check: `kubernetes_state.cronjob.complete` which returns the status of the latest job created by a cronjob.

### Motivation

The KSM core check already exposes a `kubernetes_state.job.complete` service check per job and there is a `kube_cronjob` tag attached to it to know which cronjob this job is belonging to.
But jobs are not immediately deleted after their completion so that for a given cronjob, there are multiple failed and completed jobs that are kept for a while.
Although the current `kubernetes_state.job.complete` service check can be used to know the state of every individual job, it can be valuable to know what is the status of the lastest job of a cronjob.
This wasn’t easy to get with the current service check.
See https://github.com/DataDog/integrations-core/pull/9226#issuecomment-878299670 for another description of this issue.

### Additional Notes

* Supersedes DataDog/integrations-core#9226
* Corresponding doc update: DataDog/documentation#13484
* Corresponding `integration-core` update: DataDog/integrations-core#11728

### Describe how to test/QA your changes

Create a cronjob on a kubernetes cluster.
Validate that `kubernetes_state.cronjob.complete` reflects the state of the latest job of the cronjob.

### Reviewer's Checklist

- [ ] If known, an appropriate milestone has been selected; otherwise the `Triage` milestone is set.
- [ ] The appropriate `team/..` label has been applied, if known.
- [ ] Use the `major_change` label if your change either has a major impact on the code base, is impacting multiple teams or is changing important well-established internals of the Agent. This label will be use during QA to make sure each team pay extra attention to the changed behavior. For any customer facing change use a releasenote.
- [ ] A [release note](https://github.com/DataDog/datadog-agent/blob/main/docs/dev/contributing.md#reno) has been added or the `changelog/no-changelog` label has been applied.
- [ ] Changed code has automated tests for its functionality.
- [ ] Adequate QA/testing plan information is provided if the `qa/skip-qa` label is not applied.
- [ ] If applicable, docs team has been notified or [an issue has been opened on the documentation repo](https://github.com/DataDog/documentation/issues/new).
- [ ] If applicable, the `need-change/operator` and `need-change/helm` labels have been applied.
- [ ] If applicable, the [config template](https://github.com/DataDog/datadog-agent/blob/main/pkg/config/config_template.yaml) has been updated.

### CODE REVIEW

**Review by jtappa — APPROVED** [review]
👍🏻 from docs team

File: pkg/collector/corechecks/cluster/ksm/kubernetes_state_aggregators.go [hunk 1]
```diff
@@ -106,6 +138,52 @@ func (a *countObjectsAggregator) accumulate(metric ksmstore.DDMetric) {
...
+		return
+	}
+
+	dashPosition := strings.LastIndexByte(jobName, '-')
+	if dashPosition == -1 {
+		return
+	}
+
+	cronjobName := jobName[:dashPosition]
+	cronjobID := jobName[dashPosition+1:]
```
Comment (ahmed-mez) [inline pkg/collector/corechecks/cluster/ksm/kubernetes_state_aggregators.go:173]: I was wondering if we could reuse [ParseCronJobForJob](https://github.com/DataDog/datadog-agent/blob/8419970298cae40293c6edc253a049e557decbe3/pkg/util/kubernetes/helpers.go#L46) - or maybe add a similar common helper that returns both the cronjob name and ID.

File: pkg/collector/corechecks/cluster/ksm/kubernetes_state_aggregators.go [hunk 2]
```diff
@@ -106,6 +138,52 @@ func (a *countObjectsAggregator) accumulate(metric ksmstore.DDMetric) {
...
+	id, err := strconv.Atoi(cronjobID)
+	if err != nil {
```
Comment (ahmed-mez) [inline pkg/collector/corechecks/cluster/ksm/kubernetes_state_aggregators.go:175]: a debug log here can be useful

File: pkg/collector/corechecks/cluster/ksm/kubernetes_state_aggregators.go [hunk 3]
```diff
@@ -42,6 +46,28 @@ type countObjectsAggregator struct {
...
+	name      string
+}
+
+type cronJobState struct {
+	id    int
+	state metrics.ServiceCheckStatus
+}
+
+type lastCronJobAggregator struct {
+	accumulator map[cronJob]cronJobState
```
Comment (ahmed-mez) [inline pkg/collector/corechecks/cluster/ksm/kubernetes_state_aggregators.go:60]: any reason why this is not a `map[string]cronJobState` where the key is a `ns/name` string?
I can't say using a string as key is better, it's difficult to judge without doing a benchmark.

File: pkg/collector/corechecks/cluster/ksm/kubernetes_state_aggregators.go [hunk 4]
```diff
@@ -126,6 +204,33 @@ func (a *counterAggregator) flush(sender aggregator.Sender, k *KSMCheck, labelJo
...
+				"kube_cronjob": cronjob.name,
+			},
+			labelJoiner,
+			nil,
+		)
+
+		sender.ServiceCheck(ksmMetricPrefix+"cronjob.complete", state.state, hostname, tags, "")
+	}
+
+	a.accumulator = make(map[cronJob]cronJobState)
```
Comment (ahmed-mez) [inline pkg/collector/corechecks/cluster/ksm/kubernetes_state_aggregators.go:229]: does this mean we'll send a service check on every check run even?

File: pkg/collector/corechecks/cluster/ksm/kubernetes_state_aggregators.go (hunk 1 above)
Comment (L3n41c) [reply pkg/collector/corechecks/cluster/ksm/kubernetes_state_aggregators.go]: Definitively !
Here is the change to leverage `ParseCronJobForJob`: https://github.com/DataDog/datadog-agent/pull/11139/commits/d1749e22248468ca55648a9f8188928069d99492.

File: pkg/collector/corechecks/cluster/ksm/kubernetes_state_aggregators.go (hunk 2 above)
Comment (L3n41c) [reply pkg/collector/corechecks/cluster/ksm/kubernetes_state_aggregators.go]: Debug log added: https://github.com/DataDog/datadog-agent/pull/11139/commits/d1749e22248468ca55648a9f8188928069d99492#diff-6a3196968349f588781cd7ed6e77e5136bbf45fec44ba95e257eb027de06c4aeR167

File: pkg/collector/corechecks/cluster/ksm/kubernetes_state_aggregators.go (hunk 4 above)
Comment (L3n41c) [reply pkg/collector/corechecks/cluster/ksm/kubernetes_state_aggregators.go:221]: Would it be different from how the already existing `job.complete` service check is working: https://github.com/DataDog/datadog-agent/blob/3046093abdc80c5684d304d0018acf4dfa9ee838/pkg/collector/corechecks/cluster/ksm/kubernetes_state_transformers.go#L361 ?

File: pkg/collector/corechecks/cluster/ksm/kubernetes_state_aggregators.go (hunk 3 above)
Comment (L3n41c) [reply pkg/collector/corechecks/cluster/ksm/kubernetes_state_aggregators.go:58]: It’s just that I want to index my `cronJobState` objects with two keys and it’s safer use a `struct` than building a string. Think about what would happen if `/` was a valid character in namespace and cron job names.

In addition, I’m not sure there’s a way to build the key string in a way that would be more efficient than building the `struct`.

Here is a small benchmark I did to see the order of magnitude:

```go
package main

import (
	"fmt"
	"testing"
)

var testValues = []struct {
	ns, name string
	value    int
}{
	{"nsA", "nameA", 42},
	{"nsB", "nameB", 43},
	{"nsC", "nameC", 44},
}

func Benchmark_mapStruct(b *testing.B) {

	type mapKey struct {
		namespace string
		name      string
	}

	myMap := make(map[mapKey]int)

	for n := 0; n < b.N; n++ {
		for _, val := range testValues {
			myMap[mapKey{namespace: val.ns, name: val.name}] = val.value
		}
		for _, val := range testValues {
			if v, ok := myMap[mapKey{namespace: val.ns, name: val.name}]; !ok || v != val.value {
				b.Fail()
			}
		}
	}
}

func Benchmark_mapPrintf(b *testing.B) {

	myMap := make(map[string]int)

	for n := 0; n < b.N; n++ {
		for _, val := range testValues {
			myMap[fmt.Sprintf("%s/%s", val.ns, val.name)] = val.value
		}
		for _, val := range testValues {
			if v, ok := myMap[fmt.Sprintf("%s/%s", val.ns, val.name)]; !ok || v != val.value {
				b.Fail()
			}
		}
	}
}

func Benchmark_mapStrConcat(b *testing.B) {

	myMap := make(map[string]int)

	for n := 0; n < b.N; n++ {
		for _, val := range testValues {
			myMap[fmt.Sprintf(val.ns+"/"+val.name)] = val.value
		}
		for _, val := range testValues {
			if v, ok := myMap[fmt.Sprintf(val.ns+"/"+val.name)]; !ok || v != val.value {
				b.Fail()
			}
		}
	}
}
```

```
go test -bench .
goos: darwin
goarch: amd64
pkg: map_struct
cpu: Intel(R) Core(TM) i7-8569U CPU @ 2.80GHz
Benchmark_mapStruct-8      	10415618	       115.1 ns/op
Benchmark_mapPrintf-8      	 1335531	       891.9 ns/op
Benchmark_mapStrConcat-8   	 2215784	       536.2 ns/op
PASS
ok  	map_struct	5.183s
```

Comment (ahmed-mez) [reply pkg/collector/corechecks/cluster/ksm/kubernetes_state_aggregators.go:58]: thanks for taking the time to do the benchmark 🙇

File: pkg/collector/corechecks/cluster/ksm/kubernetes_state_aggregators.go (hunk 4 above)
Comment (ahmed-mez) [reply pkg/collector/corechecks/cluster/ksm/kubernetes_state_aggregators.go:221]: I just wanted to make sure I understood correctly. Thanks for confirming!

File: pkg/collector/corechecks/cluster/ksm/kubernetes_state_aggregators.go [hunk 5]
```diff
@@ -106,6 +136,46 @@ func (a *countObjectsAggregator) accumulate(metric ksmstore.DDMetric) {
...
+	}
+
+	jobName, found := metric.Labels["job_name"]
+	if !found {
+		return
+	}
+
+	cronjobName, id := kubernetes.ParseCronJobForJob(jobName)
+	if cronjobName == "" {
+		log.Debug("%s isn't a valid CronJbo name", jobName)
```
Comment (ahmed-mez) [inline pkg/collector/corechecks/cluster/ksm/kubernetes_state_aggregators.go:167]:
```suggestion
		log.Debugf("%q isn't a valid CronJob name", jobName)
```

**Review by ahmed-mez — APPROVED** [review]

**Review by ahmed-mez — COMMENTED** [review]
Please add the new service check here https://github.com/DataDog/datadog-agent/blob/main/pkg/collector/corechecks/cluster/ksm/kubernetes_state.md#service-checks

**Comment by L3n41c** [comment]
New service check added in `kubernetes_state.md`: https://github.com/DataDog/datadog-agent/pull/11139/commits/ea26bec0d33cf9b0c8da19c58aaceea7aacdc5e8

**Pull request merged**